### PR TITLE
Generate method specs in Motoko to Viper translation

### PR DESCRIPTION
**Pull request by aterga**
This PR enables method specs in Motoko to Viper translation (including both access permissions to all actor private fields and canister invariant).

### CODE REVIEW

File: src/viper/pretty.ml [hunk 1]
```diff
@@ -117,8 +117,19 @@ and pp_exp ppf exp =
...
+  | _ -> fprintf ppf "@[// pretty printer not implemneted for node at %s@]" (string_of_region exp.at)
+
+and pp_perm ppf perm =
+  match perm.it with
+  | NoP -> fprintf ppf "none"
+  | FullP -> fprintf ppf "write"
+  | WildcardP -> fprintf ppf "wildcard"
+  | FractionalP (a, b) -> fprintf ppf "@[(%a/%a)@]" pp_exp a pp_exp b
+
+  and pp_stmt ppf stmt =
```
Comment (crusso) [inline src/viper/pretty.ml:132]:
```suggestion
and pp_stmt ppf stmt =
```

File: src/viper/pretty.ml [hunk 2]
```diff
@@ -117,8 +117,19 @@ and pp_exp ppf exp =
        | Implies _ -> "==>" | OrE _ -> "||" | AndE _ -> "&&"
        | _ -> failwith "not a binary operator" in
      fprintf ppf "(%a %s %a)" pp_exp e1 op pp_exp e2
-
-and pp_stmt ppf stmt =
+  | PermE p -> pp_perm ppf p
+  | AccE (fldacc, perm) -> fprintf ppf "@[acc(%a,%a)@]" pp_fldacc fldacc pp_exp perm
+     (* fprintf ppf "@[acc(%a.%s,%a)@]" (pp_exp exp) it (pp_exp perm) *)
```
Comment (crusso) [inline src/viper/pretty.ml:122]:
```suggestion
```

File: src/viper/pretty.ml [hunk 3]
```diff
@@ -117,8 +117,19 @@ and pp_exp ppf exp =
...
+  | _ -> fprintf ppf "@[// pretty printer not implemneted for node at %s@]" (string_of_region exp.at)
```
Comment (crusso) [inline src/viper/pretty.ml:123]:
```suggestion
  | _ -> fprintf ppf "@[// pretty printer not implemented for node at %s@]" (string_of_region exp.at)
```
(Personally, I'd prefer to assert I think)

File: src/viper/trans.ml [hunk 4]
```diff
@@ -55,7 +45,37 @@ let rec unit (u : M.comp_unit) : prog =
...
+        it = List.fold_left
+        (fun pexp -> fun p_fn -> AndE(
+            { at;
+            it = pexp;
+            note = NoInfo
+            },
+            { at;
+            it = (p_fn at);
+            note = NoInfo
+            }))
```
Comment (crusso) [inline src/viper/trans.ml:73]:
```suggestion
          { at;
            it = pexp;
            note = NoInfo
          },
          { at;
            it = (p_fn at);
            note = NoInfo
          }))
```
indentation

**Review by crusso — CHANGES_REQUESTED** [review]
Looks reasonable, but you'll need to regenerate the test output to match.

**Comment by aterga** [comment]
The nix test is failing due to 

```
error: unable to download 'https://github.com/viperproject/viperserver/releases/download/v-2022-10-13-0725/viperserver.jar': HTTP error 404
```

I suggest using [Viper July 2022 release](https://github.com/viperproject/viperserver/releases/tag/v.22.07-release) rather than nightly. The URL should then be this: 

https://github.com/viperproject/viperserver/releases/download/v.22.07-release/viperserver.jar

Wdyt, @ggreif ?

File: src/viper/trans.ml [hunk 5]
```diff
@@ -66,19 +86,37 @@ let rec unit (u : M.comp_unit) : prog =
...
+        }
+        | _ -> {it; at; note}
+    )) is'' in
+    (* Add functional invariants *)
+    let invs = extract_invariants is''' (self_id, self_typ) [] in
+    let is = List.map (fun {it; at; note: info} ->
+      match it with
+      | MethodI (id, ins, outs, pres, posts, body) ->
+        { at;
+          it = MethodI(id, ins, outs, List.append pres invs, List.append posts invs, body);
```
Comment (ggreif) [inline src/viper/trans.ml:111]: I thought that `__init__` shouldn't get the invariants as preconditions

Comment (aterga) [reply src/viper/trans.ml:111]: Good catch!

**Review by ggreif — APPROVED** [review]
LGTM!

File: src/viper/trans.ml [hunk 6]
```diff
@@ -66,19 +86,37 @@ let rec unit (u : M.comp_unit) : prog =
...
       } in
-    let is = m :: is' in
-    let invs = extract_invariants is in
-      { it = adorn_invariants invs is;
-        at = body.at;
-        note = NoInfo
-      }
+    let is'' = init_m :: is' in
+    (* Add permissions *)
+    let is''' = List.map (fun {it; at; note: info} -> (
```
Comment (ggreif) [inline src/viper/trans.ml:96]: maybe you can use `List.mapi` to observe the first element, and treat is specially w.r.t. the precondition.

File: src/viper/trans.ml [hunk 7]
```diff
@@ -66,19 +86,37 @@ let rec unit (u : M.comp_unit) : prog =
...
+      | MethodI (id, ins, outs, pres, posts, body) ->
+        { at;
+          it = MethodI (id, ins, outs, (perm at)::pres, (perm at)::posts, body);
+          note
+        }
+        | _ -> {it; at; note}
+    )) is'' in
+    (* Add functional invariants *)
+    let invs = extract_invariants is''' (self_id, self_typ) [] in
+    let is = List.map (fun {it; at; note: info} ->
```
Comment (ggreif) [inline src/viper/trans.ml:107]: maybe you can use `List.mapi` to observe the first element, and treat is specially w.r.t. the precondition.

**Pull request merged**